File: abc237/src/bin/f.rs
use proconio::input;

const MOD: usize = 998_244_353;

fn main() {
    input! {
        n: usize,
        m: usize,
    }

    let mut dp: Vec<Vec<Vec<usize>>> = vec![vec![vec![0; m + 1]; m + 1]; m + 1];
    dp[m][m][m] = 1;

    for _ in 0..n {
        let mut ndp: Vec<Vec<Vec<usize>>> = vec![vec![vec![0; m + 1]; m + 1]; m + 1];

        for a in 0..=m {
            for b in a..=m {
                for c in b..=m {
                    if dp[a][b][c] == 0 {
                        continue;
                    }

                    let d = dp[a][b][c];
                    for i in 0..m.min(c + 1) {
                        if i == c || i == b || i == c {
                            ndp[a][b][c] = (ndp[a][b][c] + d) % MOD;
                        } else if i > b {
                            ndp[a][b][i] = (ndp[a][b][i] + d) % MOD;
                        } else if i > a {
                            ndp[a][i][c] = (ndp[a][i][c] + d) % MOD;
                        } else {
                            ndp[i][b][c] = (ndp[i][b][c] + d) % MOD;
                        }
                    }
                }
            }
        }

        dp = ndp;
    }

    let mut ans = 0;
    for i in 0..m {
        for j in i + 1..m {
            for k in j + 1..m {
                ans = (ans + dp[i][j][k]) % MOD;
            }
        }
    }

    println!("{}", ans);
}

File: abc238/f/f.jl
const MOD = 998_244_353

function read_ints()::Vector{Int}
    line = readline()
    return map(x -> parse(Int, x), split(line))
end

function solve()
    n, k = read_ints()
    p = read_ints()
    q = read_ints()
    citizens = collect(zip(p, q))
    sort!(citizens)

    dp = zeros(Int, k + 1, n + 1)
    dp[1, n+1] = 1

    for rk1 in 1:n # Rk1: 1 -> n
        ndp = zeros(Int, k + 1, n + 1)

        for rk2 in 1:n+1 # Rk2 Smallest of non-selected: 1 -> n + 1
            for num in 1:k+1 # Number of people selected: 0 -> k
                if dp[num, rk2] == 0
                    continue
                end

                # Select this citizen (if possible)
                if num <= k && citizens[rk1][2] < rk2
                    ndp[num+1, rk2] += dp[num, rk2]
                    ndp[num+1, rk2] %= MOD
                end

                # Do not select this citizen
                rk2′ = min(rk2, citizens[rk1][2])
                ndp[num, rk2′] += dp[num, rk2]
                ndp[num, rk2′] %= MOD
            end
        end

        dp = ndp
    end

    ans = sum(dp[k+1, :]) % MOD
    println(ans)
end

solve()
